File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/OpenAIClient.kt
package com.intatis.shared

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.isString
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import com.intatis.shared.provider.GeneratedImage
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64
import kotlin.math.max

class OpenAIClient(private val config: IntatisConfig) {
    private val json = Json { ignoreUnknownKeys = true }

    data class ToolCall(val id: String, val name: String, val arguments: String)

    data class OpenAIChatMessage(
        val role: String,
        val content: String? = null,
        val toolCalls: List<ToolCall>? = null,
        val toolCallId: String? = null,
        val images: List<ImageAttachment> = emptyList(),
    )

    data class ToolCallResult(
        val text: String,
        val toolCalls: List<ToolCall>,
        val latencyMs: Long,
        val usage: String?,
    )

    suspend fun sendAsync(
        messages: List<IntatisMessage>,
        model: String? = null,
        reasoning: String? = null,
        attachments: List<ImageAttachment> = emptyList(),
        includeUsage: Boolean = false,
    ): Triple<String, Long, String?> {
        require(config.apiKey.isNotBlank()) { "INTATIS_API_KEY is required." }

        val resolvedModel = if (model.isNullOrBlank()) config.model else model
        val normalizedImages = attachments.filterIsInstance<ImageAttachment>()
        val request = buildMessageRequest(
            messages = messages,
            model = resolvedModel,
            reasoning = reasoning,
            images = normalizedImages,
            includeUsage = includeUsage,
        )

        val responseText = execute(request)
        val root = json.parseToJsonElement(responseText).jsonObject
        val usage = if (root["usage"] == null) null else root["usage"]!!.toString()

        val choices = root["choices"]?.jsonArray ?: throw IllegalStateException("Empty response from model provider.")
        if (choices.isEmpty()) throw IllegalStateException("Empty response from model provider.")
        val message = choices[0].jsonObject["message"]!!.jsonObject
        val content = extractContentText(message)

        val latency = 0L
        return Triple(content.trim(), latency, usage)
    }

    suspend fun sendWithToolsAsync(
        messages: List<OpenAIChatMessage>,
        tools: List<ToolDescriptor>,
        model: String? = null,
        reasoning: String? = null,
        includeUsage: Boolean = false,
    ): ToolCallResult {
        require(config.apiKey.isNotBlank()) { "INTATIS_API_KEY is required." }

        val resolvedModel = if (model.isNullOrBlank()) config.model else model
        val request = buildToolRequest(messages, tools, resolvedModel, reasoning, includeUsage)
        val responseText = execute(request)
        val root = json.parseToJsonElement(responseText).jsonObject

        val choices = root["choices"]?.jsonArray ?: throw IllegalStateException("Empty response from model provider.")
        if (choices.isEmpty()) throw IllegalStateException("Empty response from model provider.")
        val message = choices[0].jsonObject["message"]!!.jsonObject
        val content = extractContentText(message)
        val calls = parseToolCalls(message)

        val usage = if (root["usage"] == null) null else root["usage"]!!.toString()
        return ToolCallResult(content, calls, 0L, usage)
    }

    suspend fun generateImagesAsync(
        model: String,
        prompt: String,
        size: String,
        count: Int,
    ): List<GeneratedImage> {
        require(config.apiKey.isNotBlank()) { "INTATIS_API_KEY is required." }
        require(model.isNotBlank()) { "image generation model is required." }
        require(prompt.isNotBlank()) { "image prompt is required." }
        require(count in 1..4) { "image count must be in range [1, 4]." }

        val request = buildJsonObject {
            put("model", JsonPrimitive(model))
            put("prompt", JsonPrimitive(prompt))
            put("size", JsonPrimitive(size.ifBlank { "1024x1024" }))
            put("n", JsonPrimitive(count))
            put("response_format", JsonPrimitive("b64_json"))
        }

        val responseText = executeJson("/images/generations", request)
        val root = runCatching { json.parseToJsonElement(responseText).jsonObject }
            .getOrElse { throw imageGenerationResponseError(responseText, underlying = it) }

        val data = root["data"]?.jsonArray ?: throw imageGenerationResponseError(responseText)
        if (data.isEmpty()) {
            throw imageGenerationResponseError(responseText, underlying = null, detail = "provider returned no images")
        }

        return data.map { item ->
            val b64 = item.jsonObject["b64_json"]?.jsonPrimitive?.contentOrNull
                ?: throw imageGenerationResponseError(responseText, detail = "missing data[].b64_json")
            val bytes = runCatching { Base64.getDecoder().decode(b64) }
                .getOrElse { throw imageGenerationResponseError(responseText, detail = "data[].b64_json is not base64", underlying = it) }
            GeneratedImage(data = bytes, mime = "image/png")
        }
    }

    suspend fun transcribeAudioAsync(
        model: String,
        audio: ByteArray,
        filename: String,
        mimeType: String,
    ): String {
        require(config.apiKey.isNotBlank()) { "INTATIS_API_KEY is required." }
        require(model.isNotBlank()) { "transcription model is required." }
        require(filename.isNotBlank()) { "transcription filename is required." }
        require(mimeType.isNotBlank()) { "transcription mimeType is required." }
        require(audio.isNotEmpty()) { "transcription audio is required." }

        val boundary = "----IntatisBoundary${System.currentTimeMillis()}"

        val requestUrl = URL(config.baseUrl.trimEnd('/') + "/audio/transcriptions")

        val responseText = withContext(Dispatchers.IO) {
            val connection = (requestUrl.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 60_000
                readTimeout = 60_000
                doInput = true
                doOutput = true
                setRequestProperty("Authorization", "Bearer ${config.apiKey}")
                setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
                setRequestProperty("Accept", "application/json")
            }

            connection.outputStream.use { out ->
                writeMultipartFormField(out, boundary, "model", model)
                writeMultipartFileField(
                    out = out,
                    boundary = boundary,
                    fieldName = "file",
                    filename = filename.ifBlank { "audio.m4a" },
                    mimeType = mimeType,
                    audioBytes = audio,
                )
                out.write("--$boundary--\r\n".toByteArray())
            }

            val status = connection.responseCode
            val responseStream = if (status in 200..299) connection.inputStream else connection.errorStream
            val body = BufferedReader(InputStreamReader(responseStream)).use(BufferedReader::readText)
            if (status !in 200..299) {
                throw HttpException(status, body)
            }
            connection.disconnect()
            body
        }

        val root = json.parseToJsonElement(responseText).jsonObject
        val text = root["text"]?.jsonPrimitive?.contentOrNull
            ?: throw IllegalStateException("Invalid transcription response: missing text.")
        return text.trim()
    }

    suspend fun transcribeAudioAsync(
        model: String,
        audioPath: String,
        filename: String,
        mimeType: String,
    ): String {
        require(config.apiKey.isNotBlank()) { "INTATIS_API_KEY is required." }
        require(model.isNotBlank()) { "transcription model is required." }
        require(audioPath.isNotBlank()) { "transcription audio path is required." }
        require(filename.isNotBlank()) { "transcription filename is required." }
        require(mimeType.isNotBlank()) { "transcription mimeType is required." }

        val file = File(audioPath)
        require(file.exists()) { "transcription audio file not found: $audioPath" }
        require(file.isFile) { "transcription audio path must be a file: $audioPath" }

        return transcribeAudioAsync(
            model = model,
            audio = file.readBytes(),
            filename = filename.ifBlank { file.name },
            mimeType = mimeType,
        )
    }

    private suspend fun execute(payload: JsonObject): String = executeJson("/chat/completions", payload)

    private suspend fun executeJson(path: String, payload: JsonObject): String = withContext(Dispatchers.IO) {
        val requestUrl = URL(config.baseUrl.trimEnd('/') + path)
        val connection = (requestUrl.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 60_000
            readTimeout = 60_000
            doInput = true
            doOutput = true
            setRequestProperty("Authorization", "Bearer ${config.apiKey}")
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
        }

        OutputStreamWriter(connection.outputStream).use { writer ->
            writer.write(payload.toString())
            writer.flush()
        }

        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val body = BufferedReader(InputStreamReader(stream)).use(BufferedReader::readText)
        if (status !in 200..299) {
            throw HttpException(status, body)
        }
        connection.disconnect()
        body
    }

    private fun buildMessageRequest(
        messages: List<IntatisMessage>,
        model: String,
        reasoning: String?,
        images: List<ImageAttachment>,
        includeUsage: Boolean,
    ): JsonObject {
        val messagePayload = messages.mapIndexed { index, msg ->
            val msgRole = msg.role.name.lowercase()
            val messageImages = if (index == messages.lastIndex) images else emptyList()
            buildMessagePayload(msgRole, msg.content, messageImages)
        }

        return buildJsonObject {
            put("model", JsonPrimitive(model))
            put("stream", JsonPrimitive(false))
            put("messages", JsonArray(messagePayload))
            reasoning?.let { put("reasoning_effort", JsonPrimitive(it)) }
            if (includeUsage) {
                put("stream_options", buildJsonObject { put("include_usage", JsonPrimitive(true)) })
            }
        }
    }

    private fun buildToolRequest(
        messages: List<OpenAIChatMessage>,
        tools: List<ToolDescriptor>,
        model: String,
        reasoning: String?,
        includeUsage: Boolean,
    ): JsonObject {
        val payload = buildJsonObject {
            put("model", JsonPrimitive(model))
            put("stream", JsonPrimitive(false))
            put("messages", buildJsonArray {
                messages.forEach {
                    add(buildToolMessagePayload(it))
                }
            })
            put("tools", buildJsonArray {
                tools.forEach { add(it.toOpenAiDefinition()) }
            })
            put("tool_choice", JsonPrimitive("auto"))
            reasoning?.let { put("reasoning_effort", JsonPrimitive(it)) }
            if (includeUsage) {
                put("stream_options", buildJsonObject { put("include_usage", JsonPrimitive(true)) })
            }
        }
        return payload
    }

    private fun buildMessagePayload(role: String, content: String, images: List<ImageAttachment>): JsonObject {
        return if (images.isEmpty()) {
            buildJsonObject {
                put("role", JsonPrimitive(role))
                put("content", JsonPrimitive(content))
            }
        } else {
            buildJsonObject {
                put("role", JsonPrimitive(role))
                put("content", buildJsonArray {
                    if (content.isNotBlank()) {
                        add(buildJsonObject {
                            put("type", JsonPrimitive("text"))
                            put("text", JsonPrimitive(content))
                        })
                    }
                    images.forEach { image ->
                        add(buildJsonObject {
                            put("type", JsonPrimitive("image_url"))
                            put("image_url", buildJsonObject {
                                put("url", JsonPrimitive(image.url))
                            })
                        })
                    }
                })
            }
        }
    }

    private fun buildToolMessagePayload(message: OpenAIChatMessage): JsonObject {
        return buildJsonObject {
            put("role", JsonPrimitive(message.role))

            message.toolCalls?.let { calls ->
                put("tool_calls", buildJsonArray {
                    calls.forEach { call ->
                        add(
                            buildJsonObject {
                                put("id", JsonPrimitive(call.id))
                                put("type", JsonPrimitive("function"))
                                put("function", buildJsonObject {
                                    put("name", JsonPrimitive(call.name))
                                    put("arguments", JsonPrimitive(call.arguments))
                                })
                            }
                        )
                    }
                })
                put("content", buildContentPayload(message.content ?: "", emptyList()))
            } ?: run {
                put("content", buildContentPayload(message.content ?: "", message.images))
            }

            message.toolCallId?.let { put("tool_call_id", JsonPrimitive(it)) }
        }
    }

    private fun buildContentPayload(content: String, images: List<ImageAttachment>): JsonElement {
        return if (images.isEmpty()) {
            JsonPrimitive(content)
        } else {
            buildJsonArray {
                if (content.isNotBlank()) {
                    add(
                        buildJsonObject {
                            put("type", JsonPrimitive("text"))
                            put("text", JsonPrimitive(content))
                        },
                    )
                }
                images.forEach { image ->
                    add(
                        buildJsonObject {
                            put("type", JsonPrimitive("image_url"))
                            put("image_url", buildJsonObject {
                                put("url", JsonPrimitive(image.url))
                            })
                        },
                    )
                }
            }
        }
    }

    private fun extractContentText(message: JsonObject): String {
        val content = message["content"] ?: return ""
        return when {
            content is JsonPrimitive && content.isString -> content.contentOrNull ?: ""
            content is kotlinx.serialization.json.JsonArray -> {
                val pieces = content.mapNotNull { entry ->
                    val item = entry.jsonObject
                    if ((item["type"]?.jsonPrimitive?.contentOrNull ?: "") != "text") return@mapNotNull null
                    item["text"]?.jsonPrimitive?.contentOrNull
                }
                pieces.joinToString("")
            }
            else -> content.toString()
        }
    }

    private fun parseToolCalls(message: JsonObject): List<ToolCall> {
        val toolCalls = message["tool_calls"]?.jsonArray ?: return emptyList()
        return toolCalls.mapNotNull { element ->
            if (element !is JsonObject) return@mapNotNull null
            val function = element["function"]?.jsonObject ?: return@mapNotNull null
            val name = function["name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            val argsElement = function["arguments"]
            val args = when {
                argsElement == null -> "{}"
                argsElement is JsonPrimitive && argsElement.isString -> argsElement.contentOrNull ?: "{}"
                else -> argsElement.toString()
            }
            val id = element["id"]?.jsonPrimitive?.contentOrNull
                ?: "call_${System.currentTimeMillis()}"
            ToolCall(id, name, args)
        }
    }

    private fun writeMultipartFormField(
        output: java.io.OutputStream,
        boundary: String,
        name: String,
        value: String,
    ) {
        output.write("--$boundary\r\n".toByteArray())
        output.write("Content-Disposition: form-data; name=\"$name\"\r\n".toByteArray())
        output.write("Content-Type: text/plain\r\n\r\n".toByteArray())
        output.write(value.toByteArray())
        output.write("\r\n".toByteArray())
    }

    private fun writeMultipartFileField(
        output: java.io.OutputStream,
        boundary: String,
        fieldName: String,
        filename: String,
        mimeType: String,
        audioBytes: ByteArray,
    ) {
        output.write("--$boundary\r\n".toByteArray())
        output.write("Content-Disposition: form-data; name=\"$fieldName\"; filename=\"$filename\"\r\n".toByteArray())
        output.write("Content-Type: $mimeType\r\n\r\n".toByteArray())
        output.write(audioBytes)
        output.write("\r\n".toByteArray())
    }

    private fun imageGenerationResponseError(
        responseText: String,
        detail: String? = null,
        underlying: Throwable? = null,
    ): IllegalStateException {
        val parts = mutableListOf(
            "image generation returned a response that did not match OpenAI-compatible image JSON with data[].b64_json.",
            "Check endpoint compatibility, provider path, selected model, and response format.",
        )

        imageGenerationProviderMessage(responseText)?.let { parts.add("Provider said: $it") }
            ?: imageGenerationResponsePreview(responseText)?.let { parts.add("Preview: $it") }

        detail?.let { parts.add("Details: $it.") }

        underlying?.let { ex ->
            val raw = ex.message?.trim().orEmpty().take(180)
            if (raw.isNotBlank()) parts.add("Decoder said: $raw")
        }

        return IllegalStateException(parts.joinToString(" "))
    }

    private fun imageGenerationProviderMessage(responseText: String): String? {
        val root = runCatching { json.parseToJsonElement(responseText).jsonObject }
            .getOrNull() ?: return null

        val error = root["error"] as? kotlinx.serialization.json.JsonObject ?: return null
        val pieces = mutableListOf<String>()
        listOf("message", "type", "code", "param").forEach { key ->
            val value = error[key]?.jsonPrimitive?.contentOrNull
            if (!value.isNullOrBlank()) pieces.add(value)
        }
        if (pieces.isNotEmpty()) return pieces.joinToString(" ")

        for (key in listOf("message", "detail", "error_description")) {
            val value = root[key]?.jsonPrimitive?.contentOrNull
            if (!value.isNullOrBlank()) return value
        }

        return null
    }

    private fun imageGenerationResponsePreview(responseText: String): String? {
        val trimmed = responseText.trim()
        if (trimmed.isEmpty()) return null
        val compact = trimmed.replace("\n", " ").replace("\r", " ").replace("\t", " ").trim()
        return if (compact.isEmpty()) null else compact.take(180)
    }
}

class HttpException(statusCode: Int, message: String) : RuntimeException("OpenAI request failed: $statusCode. $message")

File: Intatis-Windows/src/Intatis.Shared/OpenAIClient.cs
using System.Diagnostics;
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Intatis.Windows.Shared;

public sealed class OpenAIClient
{
    private readonly IntatisConfig _config;
    private readonly HttpClient _http;

    public sealed record ToolCall(string Id, string Name, string Arguments);

    public sealed record OpenAIChatMessage(
        string Role,
        string? Content,
        IReadOnlyList<ToolCall>? ToolCalls = null,
        string? ToolCallId = null,
        IReadOnlyList<ImageAttachment>? Images = null);

    public OpenAIClient(IntatisConfig config, HttpClient? httpClient = null)
    {
        _config = config;
        _http = httpClient ?? new HttpClient();
    }

    public async Task<(string Text, TimeSpan LatencyMs, string? Usage)> SendAsync(
        IReadOnlyList<IntatisMessage> messages,
        string? model = null,
        string? reasoning = null,
        IReadOnlyList<ChatAttachment>? attachments = null,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(_config.ApiKey))
            throw new InvalidOperationException("INTATIS_API_KEY is required.");

        var sw = Stopwatch.StartNew();
        var uri = new Uri(new Uri(_config.BaseUrl.TrimEnd('/')), "/chat/completions");
        var resolvedModel = string.IsNullOrWhiteSpace(model) ? _config.Model : model;
        var normalizedImageAttachments = attachments?.OfType<ImageAttachment>().ToList() ?? new List<ImageAttachment>();

        var request = new Dictionary<string, object?>
        {
            ["model"] = resolvedModel,
            ["stream"] = false,
            ["messages"] = messages.Select((message, index) =>
            {
                IReadOnlyList<ChatAttachment>? messageAttachments = null;
                if (index == messages.Count - 1 && message.Role == MessageRole.User)
                    messageAttachments = normalizedImageAttachments;

                return BuildMessagePayload(message, messageAttachments);
            }).ToList()
        };

        if (!string.IsNullOrWhiteSpace(reasoning))
            request["reasoning_effort"] = reasoning;

        using var content = new StringContent(
            JsonSerializer.Serialize(request),
            Encoding.UTF8,
            "application/json");

        using var req = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = content
        };
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _config.ApiKey);

        using var response = await _http.SendAsync(req, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        sw.Stop();

        if (!response.IsSuccessStatusCode)
            throw new HttpRequestException($"OpenAI request failed: {(int)response.StatusCode} {response.ReasonPhrase}. {body}");

        using var doc = JsonDocument.Parse(body);
        var root = doc.RootElement;

        var usage = root.TryGetProperty("usage", out var usageEl)
            ? usageEl.GetRawText()
            : null;

        if (!root.TryGetProperty("choices", out var choices) || choices.GetArrayLength() == 0)
            throw new InvalidOperationException("Empty response from model provider.");

        var firstChoice = choices[0];
        var message = firstChoice.GetProperty("message");
        var contentText = ExtractContentText(message);
        return (contentText.Trim(), sw.Elapsed, usage);
    }

    public async Task<(string Text, IReadOnlyList<ToolCall> ToolCalls, TimeSpan LatencyMs, string? Usage)> SendWithToolsAsync(
        IReadOnlyList<OpenAIChatMessage> messages,
        IReadOnlyList<ToolDescriptor> tools,
        string? model = null,
        string? reasoning = null,
        bool includeUsage = false,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(_config.ApiKey))
            throw new InvalidOperationException("INTATIS_API_KEY is required.");

        var sw = Stopwatch.StartNew();
        var uri = new Uri(new Uri(_config.BaseUrl.TrimEnd('/')), "/chat/completions");
        var resolvedModel = string.IsNullOrWhiteSpace(model) ? _config.Model : model;

        var request = new Dictionary<string, object?>
        {
            ["model"] = resolvedModel,
            ["stream"] = false,
            ["messages"] = messages.Select(BuildMessagePayload).ToList(),
            ["tools"] = tools.Select(t => t.ToOpenAiDefinition()).ToList(),
            ["tool_choice"] = "auto",
        };

        if (!string.IsNullOrWhiteSpace(reasoning))
            request["reasoning_effort"] = reasoning;

        if (includeUsage)
            request["stream_options"] = new Dictionary<string, object?> { ["include_usage"] = true };

        using var content = new StringContent(
            JsonSerializer.Serialize(request),
            Encoding.UTF8,
            "application/json");

        using var req = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = content
        };
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _config.ApiKey);

        using var response = await _http.SendAsync(req, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        sw.Stop();

        if (!response.IsSuccessStatusCode)
            throw new HttpException(response.StatusCode, body);

        using var doc = JsonDocument.Parse(body);
        var root = doc.RootElement;

        var usage = root.TryGetProperty("usage", out var usageEl)
            ? usageEl.GetRawText()
            : null;

        if (!root.TryGetProperty("choices", out var choices) || choices.GetArrayLength() == 0)
            throw new InvalidOperationException("Empty response from model provider.");

        var firstChoice = choices[0];
        var message = firstChoice.GetProperty("message");
        var contentText = ExtractContentText(message);
        var calls = ParseToolCalls(message);
        return (contentText, calls, sw.Elapsed, usage);
    }

    private static object BuildMessagePayload(IntatisMessage message, IReadOnlyList<ChatAttachment>? attachments)
    {
        var role = message.Role.ToString().ToLowerInvariant();
        var images = attachments?.OfType<ImageAttachment>().ToList() ?? new List<ImageAttachment>();

        if (images.Length == 0)
        {
            return new Dictionary<string, object?>
            {
                ["role"] = role,
                ["content"] = message.Content
            };
        }

        var content = new List<Dictionary<string, object?>>(images.Count + 1);
        if (!string.IsNullOrWhiteSpace(message.Content))
        {
            content.Add(new Dictionary<string, object?>
            {
                ["type"] = "text",
                ["text"] = message.Content
            });
        }

        foreach (var image in images)
        {
            content.Add(new Dictionary<string, object?>
            {
                ["type"] = "image_url",
                ["image_url"] = new Dictionary<string, object?> { ["url"] = image.Url }
            });
        }

        return new Dictionary<string, object?>
        {
            ["role"] = role,
            ["content"] = content
        };
    }

    private static object BuildMessagePayload(OpenAIChatMessage message)
    {
        var payload = new Dictionary<string, object?>
        {
            ["role"] = message.Role
        };

        var content = BuildMessageContent(message.Content, message.Images);

        if (content is not null)
        {
            payload["content"] = content;
        }
        else if (message.ToolCalls is { Count: > 0 })
        {
            payload["content"] = null;
        }

        if (message.ToolCalls is { Count: > 0 })
        {
            var calls = new List<Dictionary<string, object?>>();
            for (var i = 0; i < message.ToolCalls.Count; i++)
            {
                var toolCall = message.ToolCalls[i];
                calls.Add(new Dictionary<string, object?>
                {
                    ["id"] = toolCall.Id,
                    ["type"] = "function",
                    ["function"] = new Dictionary<string, object?>
                    {
                        ["name"] = toolCall.Name,
                        ["arguments"] = toolCall.Arguments,
                    },
                });
            }
            payload["tool_calls"] = calls;
        }

        if (!string.IsNullOrWhiteSpace(message.ToolCallId))
            payload["tool_call_id"] = message.ToolCallId;

        return payload;
    }

    private static object? BuildMessageContent(string? text, IReadOnlyList<ImageAttachment>? images)
    {
        var imageList = images?.ToList() ?? new List<ImageAttachment>();
        if (imageList.Count == 0)
            return text;

        var content = new List<Dictionary<string, object?>>(imageList.Count + 1);
        if (!string.IsNullOrWhiteSpace(text))
        {
            content.Add(new Dictionary<string, object?>
            {
                ["type"] = "text",
                ["text"] = text
            });
        }

        foreach (var image in imageList)
        {
            content.Add(new Dictionary<string, object?>
            {
                ["type"] = "image_url",
                ["image_url"] = new Dictionary<string, object?> { ["url"] = image.Url }
            });
        }

        return content;
    }

    private static string ExtractContentText(JsonElement message)
    {
        if (!message.TryGetProperty("content", out var content))
            return string.Empty;

        return content.ValueKind switch
        {
            JsonValueKind.String => content.GetString() ?? string.Empty,
            JsonValueKind.Array => ExtractTextFromContentArray(content),
            _ => content.GetRawText()
        };
    }

    private static string ExtractTextFromContentArray(JsonElement content)
    {
        var output = new StringBuilder();
        foreach (var item in content.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.Object)
                continue;
            if (!item.TryGetProperty("type", out var type) || type.GetString() != "text")
                continue;
            if (item.TryGetProperty("text", out var text))
                output.Append(text.GetString());
        }
        return output.ToString();
    }

    private static List<ToolCall> ParseToolCalls(JsonElement message)
    {
        if (!message.TryGetProperty("tool_calls", out var toolCallsEl) || toolCallsEl.ValueKind != JsonValueKind.Array)
            return [];

        var result = new List<ToolCall>();
        foreach (var toolCall in toolCallsEl.EnumerateArray())
        {
            if (toolCall.ValueKind != JsonValueKind.Object)
                continue;

            if (!toolCall.TryGetProperty("function", out var functionEl) || functionEl.ValueKind != JsonValueKind.Object)
                continue;

            if (!functionEl.TryGetProperty("name", out var nameEl) ||
                nameEl.ValueKind != JsonValueKind.String)
                continue;

            var name = nameEl.GetString() ?? string.Empty;
            if (string.IsNullOrWhiteSpace(name))
                continue;

            var args = functionEl.TryGetProperty("arguments", out var argsEl)
                ? argsEl.GetRawText().Trim()
                : "{}";
            if (argsEl.ValueKind == JsonValueKind.String)
                args = argsEl.GetString() ?? "{}";

            var id = toolCall.TryGetProperty("id", out var idEl) && idEl.ValueKind == JsonValueKind.String
                ? idEl.GetString() ?? $"call_{result.Count}"
                : $"call_{result.Count}";

            result.Add(new ToolCall(id, name, args));
        }

        return result;
    }
}

public sealed class HttpException : Exception
{
    public HttpStatusCode StatusCode { get; }

    public HttpException(HttpStatusCode statusCode, string message)
        : base($"OpenAI request failed: {(int)statusCode}. {message}")
    {
        StatusCode = statusCode;
    }
}
